app/celery/broadcast_message_tasks.py

```diff
@@ -1,17 +1,93 @@
 from app.config import QueueNames
-from app.models import BroadcastEventMessageType, BroadcastProvider
-from app.dao.broadcast_message_dao import dao_get_broadcast_event_by_id, create_broadcast_provider_message
+from app.models import BroadcastEventMessageType, BroadcastProvider, BroadcastProviderMessageStatus
+from app.dao.broadcast_message_dao import (
+    dao_get_broadcast_event_by_id,
+    create_broadcast_provider_message,
+    update_broadcast_provider_message_status
+)
 
 from app.utils import format_sequential_number
 
 
+def get_retry_delay(retry_count):
+    """
+    Given a count of retries so far, return a delay for the next one.
+    `retry_count` should be 0 the first time a task fails.
+    """
+    # TODO: replace with celery's built in exponential backoff
+
+    # 2 to the power of x. 1, 2, 4, 8, 16, 32, ...
+    delay = 2**retry_count
+    # never wait longer than 5 minutes
+    return min(delay, 300)
+
+
+def check_provider_message_should_send(broadcast_event, provider):
+    """
+    If any previous event hasn't sent yet for that provider, then we shouldn't send the current event. Instead, fail and
+    raise a P1 - so that a notify team member can assess the state of the previous messages, and if necessary, can
+    replay the `send_broadcast_provider_message` task if the previous message has now been sent.
+
+    Note: This is called before the new broadcast_provider_message is created.
+
+    # Help, I've come across this code following a pagerduty alert, what should I do?
+
+    1. Find the failing broadcast_provider_message associated with the previous event that caused this to trip.
+    2. If that provider message is still failing to send, fix the issue causing that. The task to send that previous
+       message might still be retrying in the background - look for logs related to that task.
+    3. If that provider message has sent succesfully, you might need to send this task off depending on context. This
+       might not always be true though, for example, it may not be necessary to send a cancel if the original alert has
+       already expired.
+    4. If you need to re-send this task off again, you'll need to run the following command on paas:
+       `send_broadcast_provider_message.apply_async(args=(broadcast_event_id, provider), queue=QueueNames.BROADCASTS)`
+    """
+    if broadcast_event.transmitted_finishes_at < datetime.utcnow():
+        # TODO: This should be a different kind of exception to distinguish "We should know something went wrong, but
+        # no immediate action" from "We need to fix this immediately"
+        raise CBCProxyFatalException(
+            f'Cannot send broadcast_event {broadcast_event.id} ' +
+            f'to provider {provider}: ' +
+            f'The expiry time of {broadcast_event.transmitted_finishes_at} has already passed'
+        )
+
+    events = sorted(broadcast_event.broadcast_message.events, key=lambda x: x.sent_at)
```

Comment:
Optional: I had to google that this returns a list which is sorted in ascending order, so this is starting at the very first event and working forwards in time. If you were able to communicate that in a code comment or by a variable name might be useful